Sources/SPTDataLoader/SPTDataLoaderService.m

```diff
@@ -392,10 +392,17 @@ - (void)URLSession:(NSURLSession *)session
     if (handler == nil) {
         return;
     }
-    if (handler.request.backgroundPolicy == SPTDataLoaderRequestBackgroundPolicyAlways) {
-        handler.task = [session downloadTaskWithRequest:handler.request.urlRequest];
-    } else {
-        handler.task = [session dataTaskWithRequest:handler.request.urlRequest];
+
+    //
+    // This is to avoid creating a new task on an already invalidated session
+    // Usually on session invalidation all tasks are requested to be cancelled.
+    //
+    if (!([error.domain isEqualToString:NSURLErrorDomain] && error.code == NSURLErrorCancelled)) {
```

Comment:
Wouldn't be safer to use the [URLSession:didBecomeInvalidWithError:](https://developer.apple.com/documentation/foundation/nsurlsessiondelegate/1407776-urlsession) delegate to check if the session has been invalidated?

Reply:
I don't think that code is not safe. If a task is cancelled there is really no point of creating a new task that won't be used anyway. So it is either an optimization or a safeguard.
We can look into [URLSession:didBecomeInvalidWithError:](https://developer.apple.com/documentation/foundation/nsurlsessiondelegate/1407776-urlsession) as a mean to reject new requests in a follow-up PR.

Reply:
the "safe" part I mentioned was related to the comment `Usually on session invalidation all tasks are requested to be cancelled` - Is it `usually` or 100% of the times?

Reply:
There are two ways a session can be invalidated. Through `finishTasksAndInvalidate` or `invalidateAndCancel` in the former, the code I changed is not impacted. The NSURLSession instance will wait until all task are done without cancelling them in the latter instead all tasks are immediately cancelled.
I can update the comment to better clarify when the safeguard will do the job

Reply:
I have update the comment.